File: Python/Problems/Problems/problem1846.py
import heapq
import math
import sys

import numpy as np

from utils import print1DArray, print2DArray
from typing import List, Set, Optional
from collections import Counter

from .problem import Problem

class Problem1846(Problem):
    def test(self) -> bool:
        arr = [2, 2, 1, 2, 1]

        expected = 2

        result = self.maximumElementAfterDecrementingAndRearranging(arr)

        print(result)

        return result == expected

    def maximumElementAfterDecrementingAndRearranging(self, arr: List[int]) -> int:
        n = len(arr)
        arr.sort()
        th = 1
        for i in range(n):
            th = min(th + 1, arr[i])
        return th

File: C++/Problems/Problems/problem_1846.cpp
#include "problem_1846.h"

bool problem_1846::test() {
    std::vector arr { 2, 2, 1, 2, 1 };

    int expected = 2;

    auto result = maximumElementAfterDecrementingAndRearranging(arr);

    std::cout << result << std::endl;

    return result == expected;
}

int problem_1846::maximumElementAfterDecrementingAndRearranging(std::vector<int> &arr) {
    int n = arr.size();
    std::sort(arr.begin(), arr.end());
    int th = 1;
    for (int i = 1; i < n; ++i)
        th = std::min(th + 1, arr[i]);
    return th;
}
